File: python/tests/test_circular.py
"""Tests for circular measures (r_phase, edges, proj_centroid, mean_offset, etc.).

Mirror of MATLAB tests/test_circular.m.
"""
import numpy as np
import pytest
import warnings

import mpt
from mpt._utils import position_variance


class TestCircular:
    def test_balance_augmented(self):
        b = mpt.balance([0, 400, 800], None, 1200)
        assert b == pytest.approx(1.0, abs=1e-10)

    def test_balance_cluster(self):
        b = mpt.balance([0, 100, 200], None, 1200)
        assert b < 0.5  # unbalanced

    def test_evenness_whole_tone(self):
        e = mpt.evenness([0, 200, 400, 600, 800, 1000], 1200)
        assert e == pytest.approx(1.0, abs=1e-10)

    def test_coherence_diatonic(self):
        c, nc = mpt.coherence([0, 2, 4, 5, 7, 9, 11], 12)
        assert nc == 1  # one failure (tritone)
        assert c > 0.99

    def test_coherence_whole_tone(self):
        c, nc = mpt.coherence([0, 2, 4, 6, 8, 10], 12)
        assert c == pytest.approx(1.0)
        assert nc == 0

    def test_sameness_diatonic(self):
        sq, nd = mpt.sameness([0, 2, 4, 5, 7, 9, 11], 12)
        assert nd == 1
        assert sq > 0.99

    def test_sameness_whole_tone(self):
        sq, nd = mpt.sameness([0, 2, 4, 6, 8, 10], 12)
        assert sq == pytest.approx(1.0)
        assert nd == 0

    def test_edges_output_shape(self):
        e, e_signed = mpt.edges([0, 2, 4, 5, 7, 9, 11], None, 12)
        assert e.shape == (12,)
        assert np.all(e >= 0)

    def test_edges_zero_at_events_of_even_scale(self):
        """Perfect rotational symmetry ⇒ no edges at the event positions
        of a perfectly even multiset (von Mises convolution of a symmetric
        density has no derivative at the symmetry points)."""
        e, _ = mpt.edges([0, 200, 400, 600, 800, 1000], None, 1200)
        e_at_events = e[[0, 200, 400, 600, 800, 1000]]
        np.testing.assert_allclose(e_at_events, 0, atol=1e-10)

    def test_edges_signed_antisymmetric_for_block(self):
        """For a contiguous block of events, the rising-edge boundary has
        positive sign and the falling-edge boundary has negative sign."""
        # Six events filling positions 0..5 of a 12-position circle.
        _, e_signed = mpt.edges([0, 1, 2, 3, 4, 5], None, 12)
        # Rising edge expected just before position 0 (i.e., at position 11).
        # Falling edge expected just after position 5 (i.e., at position 6).
        assert e_signed[11] > 0
        assert e_signed[6] < 0

    def test_proj_centroid_balanced(self):
        y, cm, cp = mpt.proj_centroid([0, 400, 800], None, 1200)
        assert cm == pytest.approx(0.0, abs=1e-10)
        np.testing.assert_allclose(y, 0, atol=1e-10)

    def test_mean_offset_shape(self):
        h = mpt.mean_offset([0, 2, 4, 5, 7, 9, 11], None, 12)
        assert h.shape == (12,)

    def test_mean_offset_zero_at_event_positions_of_even_scale(self):
        """For an evenly spaced multiset, mean_offset is zero at each
        event position by full rotational symmetry: at any event of
        the scale, the remaining events are arranged symmetrically
        around it."""
        h = mpt.mean_offset([0, 200, 400, 600, 800, 1000], None, 1200)
        np.testing.assert_allclose(
            h[[0, 200, 400, 600, 800, 1000]], 0, atol=1e-10,
        )

    def test_circ_apm_shape(self):
        R, rp, rl = mpt.circ_apm([0, 3, 6, 10, 12], None, 16)
        assert R.shape == (16, 16)
        assert rp.shape == (16,)
        assert rl.shape == (16,)

    def test_circ_apm_autocorrelation_symmetric(self):
        """Circular autocorrelation r_lag is symmetric about lag 0:
        r_lag[k] == r_lag[period - k] for k = 1, ..., period//2."""
        _, _, r_lag = mpt.circ_apm([0, 3, 6, 10, 12], None, 16)
        for k in range(1, 9):
            assert r_lag[k] == pytest.approx(r_lag[16 - k], abs=1e-10)

    def test_markov_s_shape(self):
        y = mpt.markov_s([0, 3, 6, 10, 12], None, 16)
        assert y.shape == (16,)
        # Event positions should have positive predictions
        assert y[0] > 0
        assert y[3] > 0

    def test_markov_s_periodic_pattern_equal_at_events(self):
        """For a perfectly 4-periodic pattern in period 16, the four
        event positions share the same S-step look-ahead context, so
        the predicted weights are equal there. Non-event positions
        also share a context (with each other) and have lower weight."""
        y = mpt.markov_s([0, 4, 8, 12], None, 16)
        np.testing.assert_allclose(y[[0, 4, 8, 12]], y[0], atol=1e-10)
        # Non-events 1, 2, 3 (and their period-4 copies) form a single
        # equivalence class as well; verify they're uniform and < event y.
        np.testing.assert_allclose(y[[1, 5, 9, 13]], y[1], atol=1e-10)
        assert y[0] > y[1]

    def test_dft_circular_unison_unit_F0(self):
        """All weight at one location ⇒ |F[0]| = 1."""
        _, mag = mpt.dft_circular([100, 100, 100], None, 1200)
        assert mag[0] == pytest.approx(1.0, abs=1e-10)

    def test_dft_circular_augmented_zero_F0(self):
        """Augmented triad (cube roots of unity, scaled to the period)
        sums to zero ⇒ |F[0]| = 0."""
        _, mag = mpt.dft_circular([0, 400, 800], None, 1200)
        assert mag[0] == pytest.approx(0.0, abs=1e-10)

    def test_markov_s_shape(self):
        y = mpt.markov_s([0, 3, 6, 10, 12], None, 16)
        assert y.shape == (16,)
        # Event positions should have positive predictions
        assert y[0] > 0
        assert y[3] > 0


# ===================================================================
#  Expectation tensors
# ===================================================================

File: matlab/tests/test_circular.m
%% test_circular.m — circular measures (rPhase, edges, projCentroid, meanOffset, etc.)
%
%  Tests for circular measures (rPhase, edges, projCentroid, meanOffset, etc.).
%
%  Standalone-runnable; appends to `results` when called from
%  test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end



b = balanceCircular([0, 400, 800], [], 1200);
results{end+1,1} = 'balanceCircular: augmented triad';
results{end,2}   = abs(b - 1) < 1e-10;

b = balanceCircular([0, 100, 200], [], 1200);
results{end+1,1} = 'balanceCircular: cluster unbalanced';
results{end,2}   = b < 0.5;

e = evennessCircular([0, 200, 400, 600, 800, 1000], 1200);
results{end+1,1} = 'evennessCircular: whole-tone';
results{end,2}   = abs(e - 1) < 1e-10;

[F, mag] = dftCircular([0, 4, 7], [], 12);
results{end+1,1} = 'dftCircular: output length';
results{end,2}   = numel(F) == 3 && numel(mag) == 3;

% -- dftCircular: |F[0]| = 1 for unison --
[~, mag_uni] = dftCircular([100, 100, 100], [], 1200);
results{end+1,1} = 'dftCircular: |F[0]| = 1 for unison';
results{end,2}   = abs(mag_uni(1) - 1) < 1e-10;

% -- dftCircular: |F[0]| = 0 for augmented triad (cube roots of unity) --
[~, mag_aug] = dftCircular([0, 400, 800], [], 1200);
results{end+1,1} = 'dftCircular: |F[0]| = 0 for augmented triad';
results{end,2}   = abs(mag_aug(1)) < 1e-10;

[c, nc] = coherence([0, 2, 4, 5, 7, 9, 11], 12);
results{end+1,1} = 'coherence: diatonic nc = 1';
results{end,2}   = nc == 1;
results{end+1,1} = 'coherence: diatonic c > 0.99';
results{end,2}   = c > 0.99;

[c, nc] = coherence([0, 2, 4, 6, 8, 10], 12);
results{end+1,1} = 'coherence: whole-tone perfect';
results{end,2}   = abs(c - 1) < 1e-10 && nc == 0;

[sq, nd] = sameness([0, 2, 4, 5, 7, 9, 11], 12);
results{end+1,1} = 'sameness: diatonic nd = 1';
results{end,2}   = nd == 1;
results{end+1,1} = 'sameness: diatonic sq > 0.99';
results{end,2}   = sq > 0.99;

[sq, nd] = sameness([0, 2, 4, 6, 8, 10], 12);
results{end+1,1} = 'sameness: whole-tone perfect';
results{end,2}   = abs(sq - 1) < 1e-10 && nd == 0;

[e_out, ~] = edges([0, 2, 4, 5, 7, 9, 11], [], 12);
results{end+1,1} = 'edges: output shape and non-negative';
results{end,2}   = numel(e_out) == 12 && all(e_out >= 0);

% -- edges: zero at event positions of an even scale (rotational symmetry) --
[e_even, ~] = edges([0, 200, 400, 600, 800, 1000], [], 1200);
% MATLAB 1-indexed: positions 0, 200, 400, 600, 800, 1000 → indices 1, 201, 401, 601, 801, 1001
e_at_events = e_even([1, 201, 401, 601, 801, 1001]);
results{end+1,1} = 'edges: zero at events of even scale';
results{end,2}   = max(abs(e_at_events)) < 1e-10;

% -- edges: signed antisymmetric for a contiguous block --
% Six events filling positions 0..5 of a 12-position circle: rising edge
% just before position 0 (index 12, position 11), falling edge just
% after position 5 (index 7, position 6).
[~, e_signed] = edges([0, 1, 2, 3, 4, 5], [], 12);
results{end+1,1} = 'edges: signed > 0 at rising-edge boundary';
results{end,2}   = e_signed(12) > 0;
results{end+1,1} = 'edges: signed < 0 at falling-edge boundary';
results{end,2}   = e_signed(7) < 0;

[y, cm, ~] = projCentroid([0, 400, 800], [], 1200);
results{end+1,1} = 'projCentroid: balanced magnitude = 0';
results{end,2}   = abs(cm) < 1e-10;
results{end+1,1} = 'projCentroid: balanced projections = 0';
results{end,2}   = all(abs(y) < 1e-10);

h = meanOffset([0, 2, 4, 5, 7, 9, 11], [], 12);
results{end+1,1} = 'meanOffset: output length';
results{end,2}   = numel(h) == 12;

% -- meanOffset: zero at event positions of an even scale --
h_even = meanOffset([0, 200, 400, 600, 800, 1000], [], 1200);
h_at_events = h_even([1, 201, 401, 601, 801, 1001]);
results{end+1,1} = 'meanOffset: zero at events of even scale';
results{end,2}   = max(abs(h_at_events)) < 1e-10;

[R, rp, rl] = circApm([0, 3, 6, 10, 12], [], 16);
results{end+1,1} = 'circApm: R shape';
results{end,2}   = isequal(size(R), [16, 16]);
results{end+1,1} = 'circApm: rPhase length';
results{end,2}   = numel(rp) == 16;
results{end+1,1} = 'circApm: rLag length';
results{end,2}   = numel(rl) == 16;

% -- circApm: rLag is symmetric about lag 0
% (real-valued circular autocorrelation: r_lag(k) == r_lag(P - k))
% MATLAB 1-indexed: r_lag(k+1) == r_lag(P-k+1) for k = 1, ..., P/2 - 1.
sym_diff = max(arrayfun(@(k) abs(rl(k+1) - rl(17-k)), 1:7));
results{end+1,1} = 'circApm: rLag symmetric about lag 0';
results{end,2}   = sym_diff < 1e-10;

y = markovS([0, 3, 6, 10, 12], [], 16);
results{end+1,1} = 'markovS: output length';
results{end,2}   = numel(y) == 16;
results{end+1,1} = 'markovS: positive at events';
results{end,2}   = y(1) > 0 && y(4) > 0;

% -- markovS: equal weights for events of a 4-periodic pattern --
% In a period-16 cycle with events at 0, 4, 8, 12 (a 4-step pattern),
% the four event positions share the same S-step look-ahead context,
% so predicted weights are equal there. Same for the non-events
% within a period.
y_per = markovS([0, 4, 8, 12], [], 16);
results{end+1,1} = 'markovS: equal at events of 4-periodic pattern';
results{end,2}   = max(abs(y_per([1, 5, 9, 13]) - y_per(1))) < 1e-10;
results{end+1,1} = 'markovS: equal at non-events of 4-periodic pattern';
results{end,2}   = max(abs(y_per([2, 6, 10, 14]) - y_per(2))) < 1e-10;
results{end+1,1} = 'markovS: events outweigh non-events in periodic pattern';
results{end,2}   = y_per(1) > y_per(2);


%% ---- Standalone summary ----

if standalone
    nPass = sum([results{:, 2}]);
    nFail = size(results, 1) - nPass;
    for i = 1:size(results, 1)
        if results{i, 2}
            fprintf('  PASS  %s\n', results{i, 1});
        else
            fprintf('  FAIL  %s\n', results{i, 1});
        end
    end
    fprintf('\n=== test_circular: %d passed, %d failed (of %d) ===\n\n', ...
        nPass, nFail, nPass + nFail);
    clear cleanupDefaults
    if nFail > 0
        error('test_circular:failed', '%d test(s) failed.', nFail);
    end
end
